apps/prairielearn/src/lib/externalGrader.js

```diff
@@ -1,58 +1,73 @@
-const sqldb = require('@prairielearn/postgres');
-const externalGradingSocket = require('./externalGradingSocket');
+// @ts-check
+import * as async from 'async';
+import * as util from 'util';
+import { z } from 'zod';
+import { logger } from '@prairielearn/logger';
+import * as sqldb from '@prairielearn/postgres';
+
+import { config } from './config';
+import * as externalGradingSocket from './externalGradingSocket';
 const ExternalGraderSqs = require('./externalGraderSqs');
 const ExternalGraderLocal = require('./externalGraderLocal');
-const assessment = require('./assessment');
+import * as assessment from './assessment';
+import {
+  CourseSchema,
+  QuestionSchema,
+  GradingJobSchema,
+  SubmissionSchema,
+  VariantSchema,
+} from './db-types';
 
 const sql = sqldb.loadSqlEquiv(__filename);
 
-module.exports.init = function (callback) {
+/** @typedef Grader
+ * @property {(GradingJob, Submission, Variant, Question, Course) => import('events')} handleGradingRequest
+ */
+
+/** @type {Grader | null} */
+let grader = null;
+
+export function init() {
   if (config.externalGradingUseAws) {
     logger.verbose('External grader running on AWS');
-    module.exports.grader = new ExternalGraderSqs();
-    callback(null);
+    grader = new ExternalGraderSqs();
   } else {
     // local dev mode
     logger.verbose('External grader running locally');
-    module.exports.grader = new ExternalGraderLocal();
-    callback(null);
+    grader = new ExternalGraderLocal();
   }
-};
-
-module.exports.beginGradingJob = function (grading_job_id, callback) {
-  const params = {
-    grading_job_id,
-  };
-  sqldb.queryOneRow(sql.select_grading_job_info, params, (err, result) => {
-    if (ERR(err, callback)) return;
-    const { grading_job, submission, variant, question, course } = result.rows[0];
-
-    module.exports._beginGradingJob(grading_job, submission, variant, question, course);
-    callback(null);
-  });
-};
-
-module.exports.beginGradingJobs = function (grading_job_ids, callback) {
-  async.each(
-    grading_job_ids,
-    (grading_job_id, callback) => {
-      module.exports.beginGradingJob(grading_job_id, (err) => {
-        if (ERR(err, callback)) return;
-        callback(null);
-      });
-    },
-    (err) => {
-      if (ERR(err, callback)) return;
-      callback(null);
-    },
+}
+
+/**
+ * @param {string[]} grading_job_ids
+ */
+export async function beginGradingJobsAsync(grading_job_ids) {
+  await async.each(grading_job_ids, beginGradingJobAsync);
+}
+export const beginGradingJobs = util.callbackify(beginGradingJobsAsync);
+
+const GradingJobInfoSchema = z.object({
+  grading_job: GradingJobSchema,
+  submission: SubmissionSchema,
+  variant: VariantSchema,
+  question: QuestionSchema,
+  course: CourseSchema,
+});
+
+/**
+ *  @param {string} grading_job_id
+ */
+export async function beginGradingJobAsync(grading_job_id) {
+  if (!grader) {
+    throw new Error('External grader not initialized');
+  }
```

Comment:
For consideration: in PT we've started using `node:assert` to enforce invariants in a slightly more concise way. That would look like this:

```suggestion
  assert(grader, 'External grader not initialized');
```

(You'll need an import, of course).

Reply:
Done in d0f80be6438586844314b6eeffd40608cca934b1.